### ISSUE TITLE
Help: Frequent GPU OOM errors during model loading on custom MLServer Runtime

### ISSUE THREAD

**Issue by ericlu88** [kserve/kserve#2974]
### Discussed in https://github.com/kserve/kserve/discussions/2973

<div type='discussions-op-text'>

<sup>Originally posted by **ericlu88** June  7, 2023</sup>
Hi folks,

We are running a custom built ServingRuntime which extends from Seldon's Python MLServer. We are running LLMs like StableDiffusions and Alpaca on GPUs (A100, T4, etc). The normal inferences works fine, but what we encountered is frequent GPU OOM errors when new models are being loaded. It seems ModelMesh doesn't have visibility into the GPU side metric, but instead replies on CPU metrics (cpu load, mem) which isn't relevant in our case. We observed ModelMesh orchestrates models to already FULL pod while other empty pods are waiting in idle.

Our questions:
* Is this expected behavior? Based on this ModelMesh [issue](https://github.com/kserve/modelmesh/issues/82), it does seems ModelMesh should be able to handle it.
* If this isn't expected, how can I configure my custom runtime to honor the GPU metric?
* If this is indeed expected (i.e. not supported currently), can you share any pointers on how to extend this functionality? We are open to contribute as it seems a very commonly needed feature.

**Environment (please complete the following information):**

- OS: Ubuntu 18.04
- Version: v0.10.0

**Additional context**

Our custom MLServer ServingRuntime spec:

```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: ServingRuntime
metadata:
  name: mega-mlserver-runtime
  namespace: modelmesh-serving
spec:
  supportedModelFormats:
    - name: mega-mlserver
      version: "1"
      autoSelect: true
  multiModel: true
  grpcDataEndpoint: port:8001
  grpcEndpoint: port:8085
  replicas: 1
  containers:
    - name: mlserver
      image: <IMAGE>
      env:
        - name: MLSERVER_MODELS_DIR
          value: "/models/_mlserver_models/"
        - name: MLSERVER_GRPC_PORT
          value: "8001"
        # default value for HTTP port is 8080 which conflicts with MMesh's
        # Litelinks port
        - name: MLSERVER_HTTP_PORT
          value: "8002"
        - name: MLSERVER_LOAD_MODELS_AT_STARTUP
          value: "false"
        # Set a dummy model name via environment so that MLServer doesn't
        # error on a RepositoryIndex call when no models exist
        - name: MLSERVER_MODEL_NAME
          value: dummy-model
        # Set server addr to localhost to ensure MLServer only listen inside the pod
        - name: MLSERVER_HOST
          value: "127.0.0.1"
        # Increase gRPC max message size to support larger payloads
        # Unlimited because it will be restricted at the model mesh layer
        - name: MLSERVER_GRPC_MAX_MESSAGE_LENGTH
          value: "-1"
        - name: MLSERVER_DEBUG
          value: "true"
        - name: MLSERVER_MODEL_PARALLEL_WORKERS
          value: "4"
      resources:
        requests:
          cpu: 10
          memory: 40Gi
          nvidia.com/gpu: 1
        limits:
          cpu: "12"
          memory: 42Gi
          nvidia.com/gpu: 1
  builtInAdapter:
    serverType: mlserver
    runtimeManagementPort: 8001
    memBufferBytes: 134217728
    modelLoadingTimeoutMillis: 900000
```</div>

**Comment by yuanbai** [comment]
Hello guys,

I experienced similar issues as above. Basically I observed that ModelMesh tried to load models onto a machine that caused GPU OOMs, and then resulted failed copies.

Logs showed something like this:
`torch.cuda.OutOfMemoryError: CUDA out of memory`

Is there a way to control whether or not to load model onto a specific instance? For example:
- ModelMesh can observe what is the remaining GPU memory for a certain instance, or in another way, we explicitly specify total available GPU memory for instances.
- For models, we can configure memory request, ModelMesh can decide whether to load such a model onto an instance, by checking the instance's remaining memory vs model's required memory.

Thanks
